// ts-nameof/dist/external/transforms-ts/parse.js
"use strict";
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (Object.hasOwnProperty.call(mod, k)) result[k] = mod[k];
    result["default"] = mod;
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
var ts = __importStar(require("typescript"));
var common_1 = require("./external/common");
var common = __importStar(require("./external/transforms-common"));
var helpers_1 = require("./helpers");
var printer = ts.createPrinter({ newLine: ts.NewLineKind.LineFeed });
function parse(parsingNode, sourceFile) {
    return isNameof(parsingNode) ? parseNameof(parsingNode) : undefined;
    function parseNameof(callExpr) {
        return {
            property: parsePropertyName(callExpr),
            typeArguments: parseTypeArguments(callExpr),
            arguments: parseArguments(callExpr)
        };
    }
    function parsePropertyName(callExpr) {
        var expression = callExpr.expression;
        if (!ts.isPropertyAccessExpression(expression) || !ts.isIdentifier(expression.name))
            return undefined;
        return expression.name.text;
    }
    function parseTypeArguments(callExpr) {
        if (callExpr.typeArguments == null)
            return [];
        return callExpr.typeArguments.map(function (arg) { return parseCommonNode(arg); });
    }
    function parseArguments(callExpr) {
        return callExpr.arguments.map(function (arg) { return parseCommonNode(arg); });
    }
    function parseCommonNode(node) {
        if (ts.isPropertyAccessExpression(node))
            return parsePropertyAccessExpression(node);
        if (ts.isElementAccessExpression(node))
            return parseElementAccessExpression(node);
        if (ts.isArrowFunction(node))
            return parseFunctionReturnExpression(node, getArrowFunctionReturnExpression(node));
        if (ts.isFunctionExpression(node))
            return parseFunctionReturnExpression(node, getReturnStatementExpressionFromBlockOrThrow(node.body));
        if (ts.isNonNullExpression(node) || ts.isParenthesizedExpression(node) || ts.isAsExpression(node))
            return parseCommonNode(node.expression);
        if (ts.isQualifiedName(node))
            return parseQualifiedName(node);
        if (ts.isTypeReferenceNode(node))
            return parseCommonNode(node.typeName);
        if (ts.isSpreadElement(node))
            return parseCommonNode(node.expression);
        if (ts.isNumericLiteral(node) || helpers_1.isNegativeNumericLiteral(node))
            return parseNumeric(node);
        if (ts.isStringLiteral(node))
            return parseStringLiteral(node);
        if (ts.isArrayLiteralExpression(node))
            return parseArrayLiteralExpression(node);
        if (ts.isIdentifier(node))
            return parseIdentifier(node);
        if (ts.isImportTypeNode(node))
            return parseImportType(node);
        if (ts.isLiteralTypeNode(node))
            return parseCommonNode(node.literal);
        if (node.kind === ts.SyntaxKind.ThisKeyword)
            return common.createIdentifierNode("this");
        return common_1.throwError("Unhandled node kind (" + node.kind + ") in text: " + getNodeText(node) + " (Please open an issue if you believe this should be supported.)");
    }
    function parseArrayLiteralExpression(node) {
        var elements = node.elements.map(function (element) { return parseCommonNode(element); });
        return common.createArrayLiteralNode(elements);
    }
    function parsePropertyAccessExpression(node) {
        var expressionCommonNode = parseCommonNode(node.expression);
        var nameCommonNode = parseIdentifier(node.name);
        getEndCommonNode(expressionCommonNode).next = nameCommonNode;
        return expressionCommonNode;
    }
    function parseElementAccessExpression(node) {
        var expressionCommonNode = parseCommonNode(node.expression);
        var argumentExpressionCommonNode = parseCommonNode(node.argumentExpression);
        var computedCommonNode = common.createComputedNode(argumentExpressionCommonNode);
        getEndCommonNode(expressionCommonNode).next = computedCommonNode;
        return expressionCommonNode;
    }
    function parseQualifiedName(node) {
        var leftCommonNode = parseCommonNode(node.left);
        var rightCommonNode = parseCommonNode(node.right);
        getEndCommonNode(leftCommonNode).next = rightCommonNode;
        return leftCommonNode;
    }
    function parseNumeric(node) {
        return common.createNumericLiteralNode(getNodeValue());
        function getNodeValue() {
            if (ts.isNumericLiteral(node))
                return parseFloat(node.text);
            return helpers_1.getNegativeNumericLiteralValue(node);
        }
    }
    function parseStringLiteral(node) {
        return common.createStringLiteralNode(node.text);
    }
    function parseIdentifier(node) {
        var text = getIdentifierTextOrThrow(node);
        return common.createIdentifierNode(text);
    }
    function parseFunctionReturnExpression(functionLikeNode, node) {
        var parameterNames = functionLikeNode.parameters.map(function (p) {
            var name = p.name;
            if (ts.isIdentifier(name))
                return name.text;
            return getNodeText(name);
        });
        return common.createFunctionNode(parseCommonNode(node), parameterNames);
    }
    function parseImportType(node) {
        var importType = common.createImportTypeNode(node.isTypeOf || false, node.argument && parseCommonNode(node.argument));
        var qualifier = node.qualifier && parseCommonNode(node.qualifier);
        getEndCommonNode(importType).next = qualifier;
        return importType;
    }
    function getEndCommonNode(commonNode) {
        while (commonNode.next != null)
            commonNode = commonNode.next;
        return commonNode;
    }
    function getArrowFunctionReturnExpression(func) {
        if (ts.isBlock(func.body))
            return getReturnStatementExpressionFromBlockOrThrow(func.body);
        return func.body;
    }
    function getIdentifierTextOrThrow(node) {
        if (!ts.isIdentifier(node))
            return common_1.throwError("Expected node to be an identifier: " + getNodeText(node));
        return node.text;
    }
    function getReturnStatementExpressionFromBlockOrThrow(block) {
        return helpers_1.getReturnStatementExpressionFromBlock(block)
            || common_1.throwError("Cound not find return statement with an expression in function expression: " + getNodeText(block));
    }
    function getNodeText(node) {
        return printer.printNode(ts.EmitHint.Unspecified, node, sourceFile);
    }
    function isNameof(node) {
        if (!ts.isCallExpression(node))
            return false;
        var identifier = getIdentifierToInspect(node.expression);
        return identifier != null && identifier.text === "nameof";
        function getIdentifierToInspect(expression) {
            if (ts.isIdentifier(expression))
                return expression;
            if (ts.isPropertyAccessExpression(expression) && ts.isIdentifier(expression.expression))
                return expression.expression;
        }
    }
}
exports.parse = parse;
