// ts-nameof/dist/text/replaceInText.js
"use strict";
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (Object.hasOwnProperty.call(mod, k)) result[k] = mod[k];
    result["default"] = mod;
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
var ts = __importStar(require("typescript"));
var transforms_ts_1 = require("../external/transforms-ts");
var printer = ts.createPrinter();
function replaceInText(fileName, fileText) {
    if (arguments.length === 1) {
        fileText = fileName;
        fileName = "/file.tsx";
    }
    var sourceFile = ts.createSourceFile(fileName, fileText, ts.ScriptTarget.Latest, false);
    var transformations = [];
    var transformerFactory = function (context) {
        return function (_) { return visitSourceFile(context); };
    };
    ts.transform(sourceFile, [transformerFactory]);
    if (transformations.length === 0)
        return { replaced: false };
    return { fileText: getTransformedText(), replaced: true };
    function getTransformedText() {
        var finalText = "";
        var lastPos = 0;
        for (var _i = 0, transformations_1 = transformations; _i < transformations_1.length; _i++) {
            var transform = transformations_1[_i];
            finalText += fileText.substring(lastPos, transform.start);
            finalText += transform.text;
            lastPos = transform.end;
        }
        finalText += fileText.substring(lastPos);
        return finalText;
    }
    function visitSourceFile(context) {
        return visitNodeAndChildren(sourceFile);
        function visitNodeAndChildren(node) {
            if (node == null)
                return node;
            node = ts.visitEachChild(node, function (childNode) { return visitNodeAndChildren(childNode); }, context);
            var resultNode = transforms_ts_1.visitNode(node, sourceFile);
            var wasTransformed = resultNode !== node;
            if (wasTransformed)
                storeTransformation();
            return resultNode;
            function storeTransformation() {
                var nodeStart = node.getStart(sourceFile);
                var lastTransformation = transformations[transformations.length - 1];
                if (lastTransformation != null && lastTransformation.start > nodeStart)
                    transformations.pop();
                transformations.push({
                    start: nodeStart,
                    end: node.end,
                    text: printer.printNode(ts.EmitHint.Unspecified, resultNode, sourceFile)
                });
            }
        }
    }
}
exports.replaceInText = replaceInText;
